[coq] Allow dependencies and composition with local Coq libraries. (#2053)
ocaml/dune: doc/coq.rst

     (synopsis <text>)
     (modules <ordered_set_lang>)
     (libraries <ocaml_libraries>)
     (theories <coq_libraries>)
     (flags <coq_flags>))

> The stanza will build all ``.v`` files on the given directory. The semantics of fields is:

Reviewer: ```suggestion
The stanza will build all ``.v`` files in the given directory. The semantics of fields is:
```

- ``<module_prefix>`` is a dot-separated list of valid Coq module
  names and determines the module scope under which the theory is
  compiled [``-R`` option]. For example, if ``<module_prefix>`` is
  ``foo.Bar``, the theory modules will be named as
  ``foo.Bar.module1``, ``foo.Bar.module2``, etc... Note that modules
  in the same theory don't see the ``foo.Bar`` prefix, in the same
  way that OCaml ``wrapped`` libraries do. For compatibility reasons,

>   the 1.0 version of the Coq language installs a theory named

Reviewer: Doesn't exist yet.

Author: Thansk for the review @Zimmi48 ; I've added a note about 1.0.

  ``foo.Bar`` under ``foo/Bar``. Also note that Coq supports composing
  a module path from different theories, thus you can name a theory
  ``foo.Bar`` and a second one ``foo.Baz`` and things will work
  properly,
- the ``modules`` field allows to constraint the set of modules
  included in the theory, similarly to its OCaml counterpart. Modules
  are specified in Coq notation, that is to say ``A/b.v`` is written
  ``A.b`` in this field,
- ``public_name`` is of the form ``pkg_name.identifier``; if present,
  Dune will generate install rules for the ``.vo`` files on the
  theory. ``pkg_name`` must be a valid package name; ``identifier``

>   is not used in the 1.0 version of the language as in the legacy

Reviewer: Ditto

  install setup, all packages share a common root namespace. Indeed,
  files will be installed in ``lib/coq/user-contrib/<module_prefix>``,

>   as customary in the make-based Coq package eco-system. For

Reviewer: ```suggestion
  as customary in the make-based Coq package ecosystem. For
```

  compatibility, we also install under the ``user-contrib`` prefix the
  ``.cmxs`` files appearing in ``<ocaml_libraries>``,
- ``<coq_flags>`` will be passed to ``coqc`` as command-line options,
- the path to installed locations of ``<ocaml_libraries>`` will be passed to
  ``coqdep`` and ``coqc`` using Coq's ``-I`` flag; this allows for a Coq
  theory to depend on a ML plugin,
- your Coq theory can depend on other theories by specifying them in
  the ``<coq_libraries>`` field; if the theory is in-tree, you can use

>   either the public or a ``<module_prefix>``; Dune will then pass to

Reviewer: ```suggestion
  either the public name or a ``<module_prefix>``; Dune will then pass to
```

?

  Coq the corresponding flags to make everything compile correctly

>   [``-Q``]. For the 1.0 version of the Coq language, if the library is

Reviewer: Ditto
